bundles/org.openhab.binding.mikrotik/src/main/java/org/openhab/binding/mikrotik/internal/model/RouterosDevice.java

```diff
@@ -0,0 +1,291 @@
+            case CAP:
+                return Optional.of(new RouterosCapInterface(interfaceProps));
+            case WLAN:
+                return Optional.of(new RouterosWlanInterface(interfaceProps));
+            case PPPOE_CLIENT:
+                return Optional.of(new RouterosPPPoECliInterface(interfaceProps));
+            case L2TP_SERVER:
+                return Optional.of(new RouterosL2TPSrvInterface(interfaceProps));
+            case L2TP_CLIENT:
+                return Optional.of(new RouterosL2TPCliInterface(interfaceProps));
+            default:
+                return Optional.empty();
+        }
+    }
+
+    public RouterosDevice(String host, int port, String login, String password) {
+        this.host = host;
+        this.port = port;
+        this.login = login;
+        this.password = password;
+        this.connectionTimeout = ApiConnection.DEFAULT_CONNECTION_TIMEOUT;
+    }
+
+    public boolean isConnected() {
+        return connection != null && connection.isConnected();
+    }
+
+    public void start() throws MikrotikApiException {
+        login();
+        updateRouterboardInfo();
+    }
+
+    public void stop() {
+        if (connection != null && connection.isConnected()) {
+            logout();
+        }
+    }
+
+    public void login() throws MikrotikApiException {
+        logger.debug("Attempting login to {} ...", host);
+        this.connection = ApiConnection.connect(SocketFactory.getDefault(), host, port, connectionTimeout);
+        connection.login(login, password);
+        logger.debug("Logged in to RouterOS at {} !", host);
+    }
+
+    public void logout() {
+        logger.debug("Logging out of {}", host);
+        if (connection != null) {
+            logger.debug("Closing connection to {}", host);
+            try {
+                connection.close();
+            } catch (ApiConnectionException e) {
+                logger.debug("Logout error", e);
+            } finally {
+                connection = null;
+            }
+        }
+    }
+
+    public boolean registerForMonitoring(String interfaceName) {
+        return monitoredInterfaces.add(interfaceName);
+    }
+
+    public boolean unregisterForMonitoring(String interfaceName) {
+        return monitoredInterfaces.remove(interfaceName);
+    }
+
+    public void refresh() throws MikrotikApiException {
+        synchronized (this) {
+            updateResources();
+            updateInterfaceData();
+            updateCapsmanRegistrations();
+            updateWirelessRegistrations();
+        }
+    }
+
+    public @Nullable RouterosRouterboardInfo getRouterboardInfo() {
+        return rbInfo;
+    }
+
+    public @Nullable RouterosSystemResources getSysResources() {
+        return resourcesCache;
+    }
+
+    public @Nullable RouterosCapsmanRegistration findCapsmanRegistration(String macAddress) {
+        logger.trace("findCapsmanRegistration({}) called for {}", macAddress, host);
```

Comment:
Can this message be replaced by using the debugger? See point 4 https://www.openhab.org/docs/developer/guidelines.html#f-logging

Please check all.

Reply:
It's a `trace` which is not the default logging level and the goal is to be able to diagnose end user issues where debugger and dev env is not available.
Or you mean to say that trace logging is not encouraged in OpenHAB at all? I'm allowed myself to use it because I've seen it's usage in other bindings.

Reply:
Trace logging is not forbidden at all. But the following pattern is forbidden as it clutters the code and eats resources:
```
void myFun() {
    logger.trace("Enter myfun"); // DONT, DONT, really DONT do that
    doSomething();
    logger.trace("Leave myfun"); // DONT, DONT, really DONT do that
}
```
See the examples in the mentioned link.

If you can't reproduce a user issue and need additional logging, you could add log messages at the particular locations of interest and send the user a custom JAR file of your addon.

Reply:
Makes sense. Since Java doesn't have conditional compilation afaik, would it be acceptable to comment out trace calls?

Reply:
Actually not as it still clutters up the code. But you can stash the loggings by Git to make them available again in case you need them.

Reply:
Traces removed.

However there are plenty of them in other bindings, so the purpose is not really clear...
![image](https://user-images.githubusercontent.com/81665/126917643-105772b5-356b-461f-9db3-51fa4f9af99a.png)

Reply:
Log messages with the severity of TRACE are totally fine. What clutters the code is, if you use them for debugging.

When the program flow reaches an entry of a method, can be determined by using the debugger.